RoaringBitmap/src/main/java/org/roaringbitmap/art/Node.java

```diff
@@ -0,0 +1,353 @@
+   */
+  public static Node deserialize(ByteBuffer byteBuffer) throws IOException {
+    Node node = deserializeHeader(byteBuffer);
+    if (node != null) {
+      node.deserializeNodeBody(byteBuffer);
+      return node;
+    }
+    return null;
+  }
+
+  /**
+   * replace the node's children according to the given children parameter while doing the
+   * deserialization phase.
+   */
+  public abstract void replaceChildren(Node[] children);
+
+  /**
+   * serialize the node's body content
+   */
+  public abstract void serializeNodeBody(DataOutput dataOutput) throws IOException;
+
+  /**
+   * serialize the node's body content
+   */
+  public abstract void serializeNodeBody(ByteBuffer byteBuffer) throws IOException;
+
+  /**
+   * deserialize the node's body content
+   */
+  public abstract void deserializeNodeBody(DataInput dataInput) throws IOException;
+
+  /**
+   * deserialize the node's body content
+   */
+  public abstract void deserializeNodeBody(ByteBuffer byteBuffer) throws IOException;
+
+  /**
+   * the serialized size except the common node header part
+   *
+   * @return the size in bytes
+   */
+  public abstract int serializeNodeBodySizeInBytes();
+
+  /**
+   * insert the LeafNode as a child of the current internal node
+   *
+   * @param current current internal node
+   */
+  public static Node insertLeaf(Node current, LeafNode childNode, byte key) {
+    switch (current.nodeType) {
+      case NODE4:
+        return Node4.insert(current, childNode, key);
+      case NODE16:
+        return Node16.insert(current, childNode, key);
+      case NODE48:
+        return Node48.insert(current, childNode, key);
+      case NODE256:
+        return Node256.insert(current, childNode, key);
+      default:
+        throw new IllegalArgumentException("Not supported node type!");
+    }
+  }
+
+  /**
+   * copy the prefix between two nodes
+   */
+  public static void copyPrefix(Node src, Node dst) {
+    dst.prefixLength = src.prefixLength;
+    System.arraycopy(src.prefix, 0, dst.prefix, 0, src.prefixLength);
+  }
+
+  /**
+   * search the position of the input byte key in the node's key byte array part
+   *
+   * @param fromIndex inclusive
+   * @param toIndex exclusive
+   */
+  public static int binarySearch(byte[] key, int fromIndex, int toIndex,
+      byte k) {
+    int inputUnsignedByte = Byte.toUnsignedInt(k);
+    int low = fromIndex;
+    int high = toIndex - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      int midVal = Byte.toUnsignedInt(key[mid]);
+
+      if (midVal < inputUnsignedByte) {
+        low = mid + 1;
+      } else if (midVal > inputUnsignedByte) {
+        high = mid - 1;
+      } else {
+        return mid; // key found
+      }
+    }
+    // key not found.
+    return ILLEGAL_IDX;
+  }
+
+  private void serializeHeader(DataOutput dataOutput) throws IOException {
+    //first byte: node type
+    dataOutput.writeByte((byte) this.nodeType.ordinal());
+    //non null object count
```

Comment:
Good, I'm glad these have appropriate sizes now. 

Can empty nodes be serialised though? I think they should probably be pruned from the serialised output if they can be in the in memory representation. If empty nodes are impossible, the count can be stored in a single byte with zero mapped to 256.

Reply:
There's no empty node. The count number is in `byte` range. As I explained in the count member variable comment. It's  short type for coding benefit. As many places need to use the count, if it's byte type, you need to cast it to unsigned int before you can use the count directly.